### ISSUE TITLE
RFC: Counter initialisation

### ISSUE THREAD

**Issue by jbernardo95** [prometheus/client_ruby#141]
# Problem

The problem this proposal tries to solve is the one described [here](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics).

In essence it can be summarised as:

> Time series that are not present until something happens are difficult to deal with, as the usual simple operations are no longer sufficient to correctly handle them. 

An example of this is when we try to calculate the `increase()` of a counter from the beggining. If the first counter value is not 0 the result of calculating increase is going to be wrong: `increase([1, 1, 2, 2]) # returns 1 but should return 2 given that the counter went from 0 to 2`.

[This](https://github.com/prometheus/prometheus/issues/1853) and [this](https://github.com/prometheus/prometheus/issues/1673) are related issues.

Note that this problem does not apply to Gauges and as such I ignore them in the rest of the RFC.

# Solution

The solution to this problem is described in the Prometheus website:

> export 0 (or NaN, if 0 would be misleading) for any time series you know may exist in advance.

Note that this solution does not cover the cases where the Prometheus scrape window did not catch the counter value at 0.

For metrics without labels this is as simple as initialising the counter value to 0 and exposing it from the beggining. In fact, this is something that [other client libraries are already doing](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics).

On the other hand for metrics with several labels the only way to initialise counters for all these labels is to know the label values in advance, so that we can initialise all possible counters/label combinations. For example, if we have a counter with two labels, `app` and `env`. And the possible values for each of these labels are:

- `app`
  - `frontend`
  - `backend`
- `env`
  - `production`
  - `staging`

Then we would have to initialise the following counters to 0:

- `app=frontend,env=production`
- `app=frontend,env=staging`
- `app=backend,env=production`
- `app=backend,env=staging`

# Implementation

My proposal to solve this problem is to add a mechanism to the client that allows the user to list all possible label values upfront so that the client can initialise all the counters to 0.

Ways this could be implemented:

```ruby
# Adding yet another parameter to the metric
Counter.new(
  :http_requests_total, 
  docstring: '...', 
  labels: [:app, :env],
  label_values: {
    app: [:frontend, :backend],
    env: [:production, :staging]
  }
)

# Creating a method to provide the label values an initialise the counters
counter =
  Counter.new(
    :http_requests_total, 
    docstring: '...', 
    labels: [:app, :env]
  )

counter.init({
  app: [:frontend, :backend],
  env: [:production, :staging]
})

# Open to other suggestions
```

# Open questions

Are `label_values` or `init` good names in the alternatives above?
How does this work with Summaries and Histograms?
How does this work with `preset_labels`?
Do we want to allow the user to set the initial value to `NaN`? Or any other value?
Does this have an impact on counter resets?
Any other issues that I didn't consider?

Super keen to know your thoughts on this @brian-brazil @grobie @SuperQ @Sinjo @dmagliola.

**Comment by brian-brazil** [comment]
> Note that this problem does not apply to Gauges and as such I ignore them in the rest of the RFC.

It does.

> My proposal to solve this problem is to add a mechanism to the client that allows the user to list all possible label values upfront so that the client can initialise all the counters to 0.

We've generally left this up to users to do in their own code, but I think @beorn7 may have thoughts here. This is often something that has to be done programmatically, so can't be done in the definition.

> How does this work with Summaries and Histograms?

The same.

> How does this work with preset_labels?

As you'd expect, though I'm quite wary of that feature as it gets abused a lot and was removed from client_golang I believe.

> Do we want to allow the user to set the initial value to NaN? Or any other value?

No, that should be removed from the guidelines.

> Does this have an impact on counter resets?

None.

**Comment by beorn7** [comment]
My (still to be written down properly) plan for client_golang v2 is quite similar to your suggestion. Perhaps a bit more compact (and possibly specific to Go idioms). It will boil down to any of the vector constructors to take something called a `label.Spec` that is able to specify any number of occurrences of all of the following:
- Just a label name (which is what we have right now, possible label values are left open at this point, and users may/should initialize the zero values if they can).
- A label pair (with a fixed label value, currently implemented as `ConstLabels`).
- A label “space” (a label name with a set of possible label values).

Mock-up constructor call:
```Go
cnt := prometheus.NewCounterVec(
	"http_requests_total", "Total number of HTTP requests.",
	label.Pair("handler", "/api"), // Allows to create and register another vec for /status.
	label.Name("status"),  // Too many possibilities to specify them all.
	label.Space("version", "v1", "v2"),
)
```

The intended effect is that the vector will automatically create zero value metrics for all “fully specified” label combinations. If a vec has no “open” label names left, that will be fully automated. In the case above, once you do something like
```Go
cnt.With(label.Pairs("status", "404", "version", "v1")).Inc()
```
you will also automatically get a zero value counter for `version="v2"`.

Also, the `With` call will check that the label space is not violated. `cnt.With(label.Pairs("status", "404", "version", "v3"))` will fail.

> As you'd expect, though I'm quite wary of that feature as it gets abused a lot and was removed from client_golang I believe.

Not really. It's still useful for the creation of “curried” vecs and to add labels to a subset of metrics. We discourage from (ab-)using it for adding a label to _all_ metrics in the doc comment. Also, we have now `CurryWith` for vectors and `WrapRegistererWith` to do those things more explicitly, but they are using the ConstLabels feature under the hood. With the plan for v2, there is no explicit ConstLabels features anymore, which will hopefully help abuse, but you can have it implicitly by defining a `label.Space` with only one possible label value (and thus I decided to also allow `label.Pair` anyway).

**Comment by dmagliola** [comment]
@jbernardo95 So, this is not exactly pretty, but there's a workaround to get that behaviour...

If you `get` the value for a metric (gauge or counter or anything) for a certain set of labels, you spontaneously manifest that time series into existence.
This is not by design, it's [a side effect of how our stores work internally](https://github.com/prometheus/client_ruby/blob/master/lib/prometheus/client/data_stores/synchronized.rb#L27), but it is true for all stores provided in the official client.

This is counter-intuitive, and kind of "feels" wrong. We had a long conversation with @Sinjo about this, and we decided to leave it, mostly on account that that's how the previous version of the gem worked, so no point breaking that in case someone was counting on it, and also that you generally don't "read" your metrics in normal code. You normally just observe them / increment them / set them, and then dump them when getting scraped, so the likelihood of this being a problem was very low.

So what you could do is, after declaring the metric, `get`ting the labelsets you want to declare, and that will make the metrics show up, and that solves your problem with the current version of the client (or the one that Chris is just about to release, at least).

And what we could do, is add an `init_labelset` method for this precise purpose. Internally, it'd just call `get`, or `set(0.0)`, or whatever, but it's a way of specifying that you want this metric to appear without incrementing it. I think this is easier to understand as an interface than another (complex) parameter in the constructor.

Either way, for consistency with the existing stores, we should add to the documentation for making your own store that:
a) Reading a non-existing set of labels should return `0.0`
b) It should also create the metric so that from then on it gets reported (which is currently does for our 3 stores)

**Comment by brian-brazil** [comment]
> This is not by design

For other clients, it's be design that you can get a child that doesn't exist in order to initialise it.

**Comment by jbernardo95** [comment]
@dmagliola I wasn't aware of how things worked under the hood, that's why I suggested to add this to the constructor, but after spending some time reading through the code I really like your approach. As you said, this is as simple as setting the metric to 0 for a given set of labels.

In practice this would look like:

```ruby
counter =
  Counter.new(
    :http_requests_total, 
    docstring: '...', 
    labels: [:app, :env, :version],
    preset_labels: {version: "v1"}
  )

counter.init_labelset({
  app: ["frontend", "backend"],
  env: ["production", "staging"],
  version: ["v1", "v2"],
})

# Initialises the following time series
# http_requests_total{app="frontend", env="production", version="v1"} 0
# http_requests_total{app="frontend", env="staging", version="v1"} 0
# http_requests_total{app="backend", env="production", version="v1"} 0
# http_requests_total{app="backend", env="staging", version="v1"} 0
# http_requests_total{app="frontend", env="production", version="v2"} 0
# http_requests_total{app="frontend", env="staging", version="v2"} 0
# http_requests_total{app="backend", env="production", version="v2"} 0
# http_requests_total{app="backend", env="staging", version="v2"} 0
```

Wdyt of this approach @brian-brazil?

**Comment by brian-brazil** [comment]
That looks similar to what @beorn7  plans for the Go client, however I wouldn't have the preset_labels.

**Comment by jbernardo95** [comment]
> I wouldn't have the preset_labels.
In the constructor? This is an already existing feature, I included it in the example to exemplify how this would work with them.

**Comment by brian-brazil** [comment]
I'm not seeing how exactly these would interact, this looks like a programming error to me.

**Comment by jbernardo95** [comment]
`preset_labels` allows the user to automatically set label values.

In the example above, `counter.increment(labels: { app: "backend", env: "staging" })` would increment the `http_requests_total{app="backend", env="staging", version="v1"}` time series, because the version label is preset.

[This does not prevent the user from overriding the preset](https://github.com/prometheus/client_ruby/blob/master/lib/prometheus/client/metric.rb#L89). So the user can `counter.increment(labels: { app: "backend", env: "staging", version: "v2" })` to increment the `http_requests_total{app="backend", env="staging", version="v2"}` time series.

Going back to the example I just wanted to make it clear that we can initialise preset labels to different values. In our example initialise the version label to `v1` and `v2`. The rational for this is mainly based on the fact that you can override the `preset_labels`.

Does this make sense?

**Comment by brian-brazil** [comment]
> Does this make sense?

Overriding preset labels seems weird, you either know the label value or you don't.

**Comment by dmagliola** [comment]
```
 counter.init_labelset({
  app: ["frontend", "backend"],
  env: ["production", "staging"],
  version: ["v1", "v2"],
})
```

I wouldn't do this... The client then needs to take all combinations of all the options for each label and init those metrics. This may or may not make sense.
I'd rather you specify the actual labelsets. So:

```
counter.init_labelset(app: 'frontend', env: 'production', version: 'v1')
counter.init_labelset(app: 'frontend', env: 'production', version: 'v2')
counter.init_labelset(app: 'backend', env: 'production', version: 'v1')
counter.init_labelset(app: 'backend', env: 'production', version: 'v2')
```

Also, not sure how intentional your example labels are, but... The same instance of the same app wouldn't actually report both staging and production. You shouldn't declare the time series some other box will report (and just send 0 for them forever), in my opinion. You'll want to declare the labelsets you are going to send, so they don't spontaneously materialize with non-zero values later, but if your app will not actually send a specific labelset, it shouldn't be sent.

**Comment by franck102** [comment]
Hi all,

We have been following this and related issues for a while, and I don't think client-based approaches would solve it for us.

Our use case is AWS Lambdas pushing metrics to the push gateway; each lambda uses a unique instanceId label - so we end up with one time series per lambda.

All our range queries return wrong results because when looking at a [10m] range for example the lambdas that started executing during the range don't have a starting value.

I read the argument that Prometheus cannot know whether the missing value are because the time series is new, or a scrape was dormant for a while, or other corner cases: that argument doesn't hold w.r.t. giving users control over the behavior: we know that those metrics come from transient lambda functions, and we should be able to state that for those specific metrics a missing value should be assumed to be zero.

I don't think explicitly inserting a starting zero can fix this: if I query a one year range I expect to see increase for a lambda that started 2mn ago to be properly reported...

I am pretty new to Prometheus so I am probably not seeing the whole picture, but this issue is a serious showstopper for us.
The obvious alternative would be an aggregating gateway, but this raises other issues (different clients and native metrics for statsd for example).

**Comment by brian-brazil** [comment]
That's not an appropriate use of the pushgateway, see https://prometheus.io/docs/practices/pushing/

**Comment by franck102** [comment]
> That's not an appropriate use of the pushgateway, see https://prometheus.io/docs/practices/pushing/

As I mentioned I have researched this pretty extensively. I have read that page (3 times at least), I have looked at stats, at the weavemaker aggregating gateway, at modifying the push gateway source to do what we need...

I hear the tea is arguments, but I also see dozens or people struggling with the same issue, so at some point maybe the team will reconsider?
We for one will probably give up on Prometheus altogether if we can’t find an acceptable workaround,

**Comment by jbernardo95** [comment]
> Also, not sure how intentional your example labels are, but... The same instance of the same app wouldn't actually report both staging and production. You shouldn't declare the time series some other box will report (and just send 0 for them forever), in my opinion. You'll want to declare the labelsets you are going to send, so they don't spontaneously materialize with non-zero values later, but if your app will not actually send a specific labelset, it shouldn't be sent.

@dmagliola This is a very good point! I'll try to draft an implementation of this an will open a PR with it.

**Comment by dmagliola** [comment]
The issue originally reported is now addressed with PR #156 
@franck102 I suggest you open a separate issue for the problem you're experiencing, which is somewhat different than the one reported here.
Thank you!
